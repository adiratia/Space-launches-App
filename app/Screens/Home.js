import React, {Component } from 'react';
import {  View , SafeAreaView, ScrollView, ActivityIndicator,Alert,Platform,TouchableOpacity,Text } from 'react-native';
import {SearchBar,Button,Divider } from 'react-native-elements'
import {connect} from 'react-redux';
import { WebView } from 'react-native-webview';
import * as favoritesAction from '../store/actions/favoriteAction'
import * as wikiAction from '../store/actions/wikiAction'

import { styles } from '../style/style';
import LaunchCard from '../Common/LaunchCard';

//Home page class component include all launches
class Home extends Component {
  //Constructor
  constructor(props){
    super(props)
    this.launchesAPI = 'https://ll.thespacedevs.com/2.0.0/launch'
    //States
    this.state={
                search: '',// text from the search var
                launches:[{
                  results:[]
                }], //list of all the launches
                }
  }

  /**
   * Call before the first rendering.
   * reference : https://stackoverflow.com/questions/65662673/react-native-fetch-data-from-api-and-displaying/65662797
   */
   async componentDidMount(){

     await fetch(this.launchesAPI) // request data about 10 launches from API
      .then((response) => response.json()) //Response from request 
      .then((json) => {
        this.setState({launches:json})
      })
      .catch((error) => console.error(error)).finally(()=>{
        this.props.is_Loading(false)
      })
    }
 
  /**
   * Move to previous 10 launches page
   * If no have previuos data alert to user
   */
  prevPage = ()=>{
    if(this.state.launches.previous!=null){
    this.props.is_Loading(true)
     fetch(this.state.launches.previous) //Request data about previous 10 launches from API
    .then((response) => response.json())
    .then((json) => {
      this.setState({launches:json})
    })
    .catch((error) => console.error(error)).finally(()=>{
      this.props.is_Loading(false)
    })
  }
  else{
    //Alert  to user if on have previous data
    Alert.alert(
      "Error",
      "No previous data",
      [{ text: "OK", onPress: () => console.log("OK Pressed") }]
    );
  }
   }
 
  /**
   * Move to next 10 launches page
   * If no have more data alert to user
   */
  nextPage = ()=>{
    if(this.state.launches.next!=null){
      this.props.is_Loading(true)
      fetch(this.state.launches.next) //request data about next 10 launches from API
      .then((response) => response.json())
      .then((json) => {
        this.setState({launches:json})

      })
      .catch((error) => console.error(error)).finally(()=>{
        this.props.is_Loading(false)

      })
  }
  else{
      //Alert to user if on have data
    Alert.alert(
      "Error",
      "No more data",
      [
        { text: "OK", onPress: () => console.log("OK Pressed") }
      ]
    );
  }
   }
  
    /**
   * Search launches by name
   * @param {String} search The text from SearchBar
   */
  updateSearch = (search) => {
    this.setState({ search:search })
  };

  /**
   * Return to application from Wiki WebView
   */
  handleBackButton = () => {
    this.props.showWiki(false,this.props.wikiURL)
  };
  

    /**
 * Return WebView with wiki page
 * reference : https://heartbeat.fritz.ai/how-to-handle-navigation-with-webviews-in-a-react-native-app-1ed51ab3342f
 * @return {View}  contain WebView with wiki page
 */
renderContent() {
  return (
    <View style={{flex:1}}>
      {console.log(this.props.wikiURL)}
    <WebView
      source={{
         uri: this.props.wikiURL,
      }}
      startInLoadingState={true}
      renderLoading={() => (
        <ActivityIndicator
          color='black'
          size='large'
          style={styles.flexContainer}
        />
      )}
  /> 
    {/*Back to application button */}
    <TouchableOpacity onPress={this.handleBackButton}>
        <Text style={styles.back_button}>Back To Application</Text>
      </TouchableOpacity>
    </View>

  );
 }
  render() {
    return (
        <View style={{flex:1}}>
        {/*Check if to show the WebView */}
         {this.props.showWebView ? 
          this.renderContent()
         :
         <SafeAreaView style={styles.container}>
        {/*SearchBar to search launches by name */}
          <SearchBar
                onChangeText={this.updateSearch}
                value={this.state.search}
                containerStyle={{backgroundColor: '#f5f5f5'}}
                platform={Platform.OS}
                placeholder="Search launches..."
          />
          <View style={{ flexDirection:"row", justifyContent: 'center', alignItems: 'center' }}>
              {/*Previous and Next buttons */}
              <Button
                style={styles.btn}
                title="<< Previous"
                type="outline"
                onPress = {this.prevPage}
              />
              {/*Space between the buttons */}
              <View style={styles.space} /> 
              <Button
                  style={styles.btn}
                  title="Next >>"
                  type="outline"
                  onPress= {this.nextPage}
              />

            </View>
            <ScrollView style ={{width:'100%'}} >
            {
              /*
               * Check if page is loading 
               * if yes show a spinner
               * else show launches list
              */
            this.props.isLoading ? <ActivityIndicator size="large"  />
            :
            /* Check  if the launches exist */
            typeof(this.state.launches.detail)!=='undefined'  ? 
            <View>
              {            console.log(this.state.launches.detail)
}
               { Alert.alert(
              "Error",
              "Request was throttled",
              [{ text: "OK", onPress: () => console.log("OK Pressed") }]
            )}
            </View>
            :
          this.state.launches?.results?.map((u, i) => {
                /*
                 * Check if the search bar empty or the length of the text less 3 characters then return all launches cards 
                 * else return only the cards in the currnet page that contain the text from search bar
                */         
               if(this.state.search=== '' || this.state.search.length <3  ){
              return (       
                <View key={i} >
                    {
                  /**Create a card for each launch
                   * Cards style taken from : https://github.com/SiDevesh/React-Native-Material-Cards
                   */
                    }
                    <LaunchCard name={u.name} image={u.image} country={u.pad.location.name} date={u.net} url={u.rocket.configuration.url} status={u.status.name} />
                    <Divider orientation="horizontal" />
                </View>
              );
              }
              else{
                /*Show only the launches cards that includes the search box text */
                if( u.name.toLowerCase().includes(this.state.search.toLowerCase())){
                  return ( 
                    <View key={i} >
                        {/*Create a card for each launch*/}
                        <LaunchCard name={u.name} image={u.image} country={u.pad.location.name} date={u.net} url={u.rocket.configuration.url} status={u.status.name} />   
                    </View>
    
                  );
                }
              }      
            }
            )

          }    
     </ScrollView>
    </SafeAreaView>
        }
        </View>
    );
  }
}
/**
 * Map the global favorites state to props
*/
const mapStateToProps = state =>{
    return {
        ...state.favorites,
        ...state.wiki

    }
}
/**
 * Map the dispatch to props
*/
const mapDispatchToProps =dispatch =>{
    return {
      addToFavorites:(name,image,date,status,country,wiki)=>{
        dispatch(favoritesAction.addToFavorites(name,image,date,status,country,wiki));
    },
    removeFromFavorites:(name)=>{
      dispatch(favoritesAction.removeFromFavorites(name));
  },
    showWiki: (showWebView,wikiURL) =>{
      dispatch(wikiAction.showWiki(showWebView,wikiURL));

    },
    is_Loading : (isLoading) =>{
      dispatch(wikiAction.is_Loading(isLoading))
    }
}
}
export default connect(
  mapStateToProps,
  mapDispatchToProps)(Home);