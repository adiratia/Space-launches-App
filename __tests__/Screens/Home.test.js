import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import Home from '../../app/Screens/Home'
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';

jest.useFakeTimers();


const mockStore = configureMockStore([]);
 
describe('Connected to Home Component', () => {
  let store;
  let component;

 
  beforeEach(() => {
    store = mockStore({
    });
  store.dispatch = jest.fn();
 
  component = renderer.create(
    <Provider store={store}>
      <Home />
    </Provider>
  );
  });
 
  it('should render from Redux store', () => {
    renderer.act(() => {
        expect(component.toJSON()).toMatchSnapshot();
    });
  });
 
});
