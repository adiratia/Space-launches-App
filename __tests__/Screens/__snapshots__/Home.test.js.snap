// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected to Home Component should render from Redux store 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5f5f5",
        "flex": 1,
        "justifyContent": "center",
        "paddingTop": null,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f5f5f5",
          "flexDirection": "row",
          "overflow": "hidden",
          "paddingBottom": 13,
          "paddingTop": 13,
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#dcdce1",
              "borderBottomWidth": 0,
              "borderColor": "#86939e",
              "borderRadius": 9,
              "flexDirection": "row",
              "marginLeft": 8,
              "marginRight": 8,
              "minHeight": 36,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onBlur={[Function]}
            onCancel={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Search launches..."
            placeholderTextColor="#7d7d7d"
            platform="ios"
            rejectResponderTermination={true}
            style={
              Object {
                "color": "#242424",
                "flex": 1,
                "fontSize": 18,
                "marginLeft": 6,
                "minHeight": 40,
                "overflow": "hidden",
              }
            }
            testID="searchInput"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginRight": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            />
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "opacity": 0,
            "position": "absolute",
            "right": null,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007aff",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": 60,
              "opacity": 1,
              "width": 120,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0.5,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2089dc",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              &lt;&lt; Previous
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": 60,
              "opacity": 1,
              "width": 120,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0.5,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2089dc",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Next &gt;&gt;
            </Text>
          </View>
        </View>
      </View>
    </View>
    <RCTScrollView
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <View />
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
